src/diff_xdiff.c

```diff
@@ -115,15 +116,19 @@ static int git_xdiff_cb(void *priv, mmbuffer_t *bufs, int len)
 	const git_diff_delta *delta = patch->base.delta;
 	git_patch_generated_output *output = &info->xo->output;
 	git_diff_line line;
+	int buffer_len;
 
 	if (len == 1) {
 		output->error = git_xdiff_parse_hunk(&info->hunk, bufs[0].ptr);
 		if (output->error < 0)
 			return output->error;
 
-		info->hunk.header_len = bufs[0].size;
 		if (info->hunk.header_len >= sizeof(info->hunk.header))
 			info->hunk.header_len = sizeof(info->hunk.header) - 1;
+
+		/* Sanitize the hunk header in case there is invalid Unicode */
+		buffer_len = git__utf8_valid_buf_length((const uint8_t *) bufs[0].ptr, bufs[0].size);
+		info->hunk.header_len = buffer_len;
```

Comment:
On that note, I don't understand the _test_ then.  Did you produce a file that has invalid utf-8 characters on a line, and are ensuring that they're truncated?  I ask because it appears that the test is doing _something_, since it appears that your `git__utf8_valid_buf_length` function has determined that the last two bytes do not include valid bytes to make up a utf-8 character.

I expected that the test would have done the truncation at a point that produced an some bytes that were invalid UTF-8 and then truncated them.  I think _that_ test would have illustrated this problem, since there would have been no truncation at all.

But I may be misunderstanding the test...?